### PR TITLE
[ticket/14703] Fix parent module selection for custom extension modules

### PR DESCRIPTION
**Pull request by rxu**
The PR is for 3.1.x, 3.2.x and master.

Checklist:

- [x] Correct branch: master for new features; 3.2.x, 3.1.x for fixes
- [x] Tests pass
- [x] Code follows coding guidelines: [master / 3.2.x](https://area51.phpbb.com/docs/master/coding-guidelines.html), [3.1.x](https://area51.phpbb.com/docs/31x/coding-guidelines.html)
- [x] Commit follows commit message [format](https://wiki.phpbb.com/Git#Commit_Messages)

<a href="https://tracker.phpbb.com/browse/PHPBB3-14703">PHPBB3-14703</a>.

### CODE REVIEW

File: phpBB/phpbb/db/migration/tool/module.php [hunk 1]
```diff
@@ -210,12 +210,15 @@ public function add($class, $parent = 0, $data = array())
...
 		// The "manual" way
+		// More than 1 module with the same module_basename may exist
+		// Thus use empty module_basename to select a category as a parent
 		if (!is_numeric($parent))
 		{
 			$sql = 'SELECT module_id
 				FROM ' . $this->modules_table . "
 				WHERE module_langname = '" . $this->db->sql_escape($parent) . "'
-					AND module_class = '" . $this->db->sql_escape($class) . "'";
+					AND module_class = '" . $this->db->sql_escape($class) . "';
```
Comment (VSEphpbb) [inline phpBB/phpbb/db/migration/tool/module.php]: semi colon at end should be concatenating

**Comment by VSEphpbb** [comment]
Not sure this will work. for example, ACP_ATTACHMENT_SETTINGS appears twice and both times have a module_basename with data.

**Comment by rxu** [comment]
ACP_ATTACHMENTS is the parent category, ACP_ATTACHMENT_SETTINGS is an actual module and shouldn't be parent whenever as far as I see it.

**Comment by VSEphpbb** [comment]
I'm just saying I'm not 100% sure that the desired target module will always have an empty basename, which this change might be forcing to be the case.

**Comment by rxu** [comment]
Well, this is indeed the case for UCP_PM category which has basename set (```ucp_pm```).
Then we need a method to extract categories from the modules table, but it's not clear how categories differ from real modules other than by basenames.

**Comment by rxu** [comment]
Ok. For the start, the 1st-level categories have parent_id of 0 which makes it easier to select. Thus, to get the 2nd level cats, we could select ones which parents has parent_id of 0.

**Comment by rxu** [comment]
!set WIP

**Comment by rxu** [comment]
!unset WIP

**Comment by rxu** [comment]
Updated. For the reference, tests already exist here https://github.com/phpbb/phpbb/blob/3.1.x/tests/dbal/migrator_tool_module_test.php

**Comment by rxu** [comment]
!set WIP

**Comment by marc1706** [comment]
Is the WIP label still correct?

**Comment by rxu** [comment]
Yes, I'm planning to implement getting parent id when before/after module parameters are set.

**Comment by marc1706** [comment]
Ok, thanks for the update.

**Comment by rxu** [comment]
Finally it works.
The patch may seem hackish, but as long as the ```module_langname``` is not unigue, workarounding is the only way I see here.

**Comment by rxu** [comment]
!unset WIP

File: tests/dbal/migrator_tool_module_test.php [hunk 2]
```diff
@@ -118,6 +118,40 @@ public function test_add()
...
+		{
+			$this->tool->add('acp', 'ACP_FORUM_BASED_PERMISSIONS', array(
+				'module_basename'	=> 'acp_new_permissions_module',
+				'module_langname'	=> 'ACP_NEW_PERMISSIONS_MODULE',
+				'module_mode'		=> 'test',
+				'module_auth'		=> '',
+			));
+			$this->fail('Exception not thrown');
+		}
+		catch (Exception $e) {}
```
Comment (marc1706) [inline tests/dbal/migrator_tool_module_test.php]: You should probably use setExpectedException() to test this

Comment (marc1706) [inline tests/dbal/migrator_tool_module_test.php]: Please do check the actual contents of the exception (the message for example) though

File: phpBB/phpbb/db/migration/tool/module.php [hunk 3]
```diff
@@ -487,4 +432,110 @@ protected function get_module_info($class, $basename)
...
+							WHERE module_langname = '" . $this->db->sql_escape($neighbour_module_langname) . "'
+								AND " . $this->db->sql_in_set('parent_id', $ids);
+						$result = $this->db->sql_query($sql);
+						$parent_id = (int) $this->db->sql_fetchfield('parent_id');
+						if (!$parent_id)
+						{
+							throw new \phpbb\db\migration\exception('PARENT_MODULE_FIND_ERROR', $data['parent_id']);
+						}
+					}
+					else if (!empty($data) && !is_array($data))
```
Comment (marc1706) [inline phpBB/phpbb/db/migration/tool/module.php]: it seems like this part wasn't covered in tests?

File: tests/dbal/migrator_tool_module_test.php (hunk 2 above)
Comment (rxu) [inline tests/dbal/migrator_tool_module_test.php]: I have no idea how to do it.

Comment (marc1706) [inline tests/dbal/migrator_tool_module_test.php]: You should be able to check the contents and type of the exception, right?

File: phpBB/phpbb/db/migration/tool/module.php (hunk 3 above)
Comment (marc1706) [inline phpBB/phpbb/db/migration/tool/module.php]: Bump.

File: tests/dbal/migrator_tool_module_test.php (hunk 2 above)
Comment (rxu) [inline tests/dbal/migrator_tool_module_test.php]: Is it something like
```php
catch (Exception $e) {
	$this->assertInstanceOf('exception', get_class($e));
	$this->assertEquals('PARENT_MODULE_FIND_ERROR', $e->getMessage());
}
```

File: phpBB/phpbb/db/migration/tool/module.php (hunk 3 above)
Comment (rxu) [inline phpBB/phpbb/db/migration/tool/module.php]: It should be covered by ```test_exists()``` as this part of code is involved for a module existance checking.

**Comment by rxu** [comment]
Updated #4377.

**Comment by Nicofuma** [comment]
LGTM @marc1706

**Pull request merged**